### PR TITLE
feat: Add DeepSeek V4 SWA recompute

### PR DESCRIPTION
**Pull request by jellysnack**
## Motivation

DeepSeek V4 keeps a FULL cache containing compressed C4/C128 and indexer state, plus a sliding-window attention (SWA) cache backed by a smaller KV pool. When this pool evicts the trailing SWA pages while the FULL prefix survives, the prefix cannot be reused directly and may fall back to cold prefill. This PR retains the FULL hit and recomputes only the bounded trailing dependency window needed to rebuild SWA.

The feature is disabled by default and enabled with `SGLANG_OPT_SWA_RECOMPUTE_WINDOW=1`.

### Cache matching and scheduling

- Compute a page-aligned recompute window from the sliding window `W`, number of layers `N`, and KV page size:
  `W_r = align_up(align_up(W, page_size) + (N - 1) * W, page_size)`.
- Trigger recompute only when the trailing SWA data is missing from both GPU and host and the matched FULL prefix reaches a configurable threshold (`2 * W_r` by default). Otherwise, use normal cache matching or HiCache load-back.
- Report the recompute length as `extra_compute_prefix_len` so the scheduler includes it in prefill admission, chunking, and SWA KV capacity checks.
- Keep the matched prefix under a FULL-only lock and rebuild SWA in temporary private pages, leaving the global mapping unchanged during the forward. On success, publish and lock only the final SWA tail and free the remaining workspace. On abort, discard all private pages.

### DeepSeek V4 execution

- Replay the selected trailing prefix together with new input tokens while preserving the original logical prefix boundary in DeepSeek V4 metadata.
- Pass that boundary to Compressor V2 so replayed prefix positions skip duplicate FULL-side compression work while still producing the SWA state needed for recompute.
- Route SWA reads and writes for replayed tokens through private pages in both prefill planner stages and in dense/sparse attention. Positions outside the replayed range keep the normal FULL-to-SWA mapping.

### HiCache

- Allow eligible Mooncake L3 hits to omit trailing SWA pages. Fall back to cold prefill if the loaded state is insufficient for recompute.

### Current limitations and guards

- The supported path is hybrid-SWA DeepSeek V4 on CUDA with `UnifiedRadixCache` and Compressor V2.
- HIP, NPU, XPU, speculative decoding, online compression, decode context parallelism, and `unified_kv_triton` are rejected.
- HiCache L3 supports Mooncake only.
- Recompute requests do not support input embeddings, positional embedding overrides, or prompt/input logprobs. Output logprobs remain supported.
- Prefill CUDA graphs and two-batch overlap are disabled for batches that contain recompute work.
- `--chunked-prefill-size` must be at least `W_r + page_size`.

## Validation

Registered tests cover window sizing, cache matching and fallback, private-page commit/abort, scheduler accounting, DeepSeek V4 prefill paths, Mooncake dispatch, and configuration guards.

## Performance

Tested on `sgl-project/DeepSeek-V4-Flash-FP8` using NVIDIA H200 GPUs. To isolate prefill cost, the benchmark uses fake-prefill transfer while keeping cache insertion enabled. SWA capacity is constrained with `--swa-full-tokens-ratio 0.005`.

The test sends a 128K-token request with `seed=42`, uses `seed=43` to evict its SWA tail, and then repeats `seed=42`.

<details>
<summary>Benchmark command and cache sizes</summary>

```bash
python3 -m sglang.bench_serving \
  --base-url http://127.0.0.1:30000 \
  --backend sglang \
  --model sgl-project/DeepSeek-V4-Flash-FP8 \
  --num-prompts 1 \
  --dataset-name random \
  --random-input-len 128000 \
  --random-output-len 1 \
  --random-range-ratio 1 \
  --warmup-requests 0 \
  --seed 42
```

```text
DSV4 pool sizes: full=7065856, swa=35328, c4=1766464, c128=55202, c4_state=2208, c128_state=0
```

</details>

| Configuration | Request | Input throughput (tok/s) | TTFT (ms) |
| --- | --- | ---: | ---: |
| Recompute disabled | Cold | 44,622 | 2,856 |
| Recompute disabled | Post-eviction | 45,140 | 2,824 |
| Recompute enabled | Cold | 44,864 | 2,843 |
| Recompute enabled | Post-eviction | 212,741 | 589 |

Without recompute, the repeated request remains at cold-prefill performance. With recompute, it reuses 122,112 FULL tokens and prefills only `W_r + page_size = 5,888` tokens. Compared with the feature-disabled post-eviction request, throughput improves by `4.71x`, TTFT improves by `4.79x`, and physical prefill work drops by `95.4%`.

### Recompute length gate

For `sgl-project/DeepSeek-V4-Flash-FP8`, the physical recompute range is 5,888 tokens. Measurements around this size show that, for short prefixes, the saved prefill work does not outweigh the recompute overhead:

| Input tokens | Cached FULL tokens | Cold TTFT (ms) | Recompute TTFT (ms) | TTFT change |
| ---: | ---: | ---: | ---: | ---: |
| 6,144 (~1x) | 256 | 232 | 334 | 44.0% slower |
| 8,960 (~1.5x) | 3,072 | 224 | 230 | 2.7% slower |
| 11,776 (2x) | 5,888 | 280 | 243 | 13.2% faster |
| 17,664 (3x) | 11,776 | 538 | 251 | 53.3% faster |

The crossover is between approximately `1.5x` and `2x`, which motivates `2.0` as the default for `SGLANG_SWA_RECOMPUTE_GATE_MULTIPLIER`. The value remains configurable because the crossover depends on hardware, batch shape, and serving load.

#### Default-path regression check

With recompute disabled, gsm8k was run on `sgl-project/DeepSeek-V4-Flash-FP8` using NVIDIA H200 GPUs.

```
python3 gsm8k/bench_sglang.py --num-shots 8 --num-questions 1209 --parallel 32 --platinum
Accuracy: 0.978
```

#### BFCL multi-turn long-context

Prompt/input logits are not exposed for internally recomputed tokens, so an exact logit comparison is unavailable. We use BFCL `multi_turn_long_context` as an end-to-end proxy because tool results repeatedly extend the same conversation prefix and produce frequent cache hits.

SWA eviction depends on cache pressure and request timing, so a normal benchmark run cannot guarantee enough recompute events. We therefore collected 50 baseline measurements and 50 with `SGLANG_DEBUG_FORCE_SWA_RECOMPUTE=1`, which forces eligible cache hits through recompute:

| Mode | Runs | Mean | Std | 95% bootstrap CI of mean |
|---|---:|---:|---:|---:|
| Baseline | 50 | 49.81 | 2.32 | [49.17, 50.45] |
| Force recompute | 50 | 50.36 | 2.33 | [49.73, 51.01] |

The mean difference is `+0.55` BFCL points, with a 95% bootstrap CI of `[-0.36, +1.45]`. Welch's t-test finds no statistically detectable difference (`p = 0.24`).

---
### CI States

Latest PR Test (Base): :x: [Run #29687292774](https://github.com/sgl-project/sglang/actions/runs/29687292774)
Latest PR Test (Extra): :x: [Run #29687292668](https://github.com/sgl-project/sglang/actions/runs/29687292668)

### CODE REVIEW

File: python/sglang/srt/mem_cache/swa_recompute.py [hunk 1]
```diff
@@ -0,0 +1,410 @@
...
+                            [seq_len], dtype=torch.int64, device=batch.device
+                        ),
+                        seq_lens_cpu=torch.tensor([seq_len], dtype=torch.int64),
+                        last_loc=last_loc,
+                        extend_num_tokens=new_tokens,
+                    )
+                batch.req_to_token_pool.req_to_token[req_idx, prefix_len:seq_len] = tail
+                out_cache_parts.append(
+                    torch.cat([recompute_full, tail]) if recompute_len > 0 else tail
+                )
```
[the remark this answers is not in the document]
Comment (jellysnack) [reply python/sglang/srt/mem_cache/swa_recompute.py:269]: `Req.prefix_indices` is a `torch.Tensor`. It is explicitly [declared and initialized as `torch.Tensor`](https://github.com/jellysnack/sglang/blob/7a03d3014935af5540482941a8c5240695beb29e/python/sglang/srt/managers/schedule_batch.py#L891), and prefix matching preserves this contract by assigning `MatchResult.device_indices`, which is also a tensor

File: python/sglang/srt/mem_cache/swa_recompute.py [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+                    assert req.swa_prefix_lock_released, (
+                        "SWA-window recompute requires a FULL-only request lock "
+                        f"before allocating private SWA pages, rid={req.rid}"
+                    )
+                    full_indices = req.prefix_indices[
+                        prefix_len - recompute_len : prefix_len
+                    ]
+                    fresh_swa = allocator.alloc_fresh_swa_for_recompute_window(
+                        full_indices
+                    )
```
[the remark this answers is not in the document]
Comment (jellysnack) [reply python/sglang/srt/mem_cache/swa_recompute.py:313]: `Req.prefix_indices` is explicitly [declared and initialized as a `torch.Tensor`](https://github.com/jellysnack/sglang/blob/7a03d3014935af5540482941a8c5240695beb29e/python/sglang/srt/managers/schedule_batch.py#L891). Therefore slicing it produces a tensor
